***Solving a linear system Ax = b with (random) 7000 × 7000 matrix A***

In [28]:
import numpy as np
import timeit

n = 7000  
A = np.random.rand(n, n)
b = np.random.rand(n, 1)

def solve_linear_system():
    return np.linalg.solve(A, b)

execution_time = timeit.timeit(solve_linear_system, number=70) 
print(f"Python Linear System Mean Time: {execution_time / 70} seconds")

Python Linear System Solution Time: 4.635231105714254 seconds


first attempt: Python Linear System Mean Time: 4.4650140600000405 seconds

second attempt: Python Linear System Mean Time: 4.437492304285739 seconds

***Computing the eigenvalues/vectors of a (random) 1200 × 1200 matrix A***

In [42]:
import numpy as np
import timeit

np.random.seed(123)

n = 1200 

A = np.random.rand(n, n)

def compute_eigen():
    eigenvalues, eigenvectors = np.linalg.eig(A)

mean_time = timeit.timeit(compute_eigen, number=70) / 70

print(f"NumPy Eigenvalues and Eigenvectors Mean Time: {mean_time} seconds")

NumPy Eigenvalues and Eigenvectors Mean Time: 2.978925727142736 seconds


first attempt: NumPy Eigenvalues and Eigenvectors Mean Time: 2.9791387414285313 seconds

second attempt: NumPy Eigenvalues and Eigenvectors Mean Time: 2.978925727142736 seconds

***Computing the singular values/vectors of a (random) 2100 × 2100
matrix A***

In [53]:
import numpy as np
import timeit

n = 2100

A = np.random.randn(n, n)

num_iterations = 70

mean_time = timeit.timeit(lambda: np.linalg.svd(A), number=num_iterations) / num_iterations

print("Mean time for Singular Value Decomposition: {} seconds".format(mean_time))

Mean time for Singular Value Decomposition: 3.7360818928569124 seconds


first attempt: Mean time for Singular Value Decomposition: 4.13997880285689 seconds

second attempt: Mean time for Singular Value Decomposition: 3.7360818928569124 seconds